lib/roken/resolve.c

```diff
@@ -690,7 +696,9 @@ rk_dns_srv_order(struct rk_dns_reply *r)
 	    (*tt)->next = *headp;
 	    *headp = *tt;
 	    headp = &(*tt)->next;
-	    sum -= (*tt)->u.srv->weight;
+	    sum -= (*tt)->u.srv->weight * zeros;
+	    if ((*tt)->u.srv->weight == 0)
+		sum--;
 	    *tt = NULL;
```

Comment:
This (*tt = NULL) part had me confused for a bit...  While it is not new with this patch and should be improved separately, for the record instead of filling the array with NULLs to skip, and adding code to skip NULLs, ... it would surely be much simpler to swap the selected element with (*ss) and then increment ss.  May well do that, when I get a chance...